s3fs/core.py

```diff
@@ -645,10 +651,11 @@ async def _lsdir(
                 dircache = []
                 async for i in it:
                     dircache.extend(i.get("CommonPrefixes", []))
-                    for c in i.get("Contents", []):
-                        c["type"] = "file"
-                        c["size"] = c["Size"]
-                        files.append(c)
+                    for c in i.get(contents_key, []):
+                        if not self.version_aware or c.get("IsLatest"):
```

Comment:
So the only potential downside I see, is that this call might take substantially longer. Is there a way to specify that we do wand the version info, but only for the latest entry at each key?

Reply:
Unfortunately there is not. `list_objects_v2` always returns the latest entry, but does not include version IDs. `list_object_versions` does not allow you to specify that you only want the latest versions.

We could pass a jmespath query into the paginator to filter results by the `IsLatest` key, but that's still client-side filtering and doesn't actually help us avoid needing to get all of the versions from S3.

Reply:
So for the user who doesn't want this problem and knows they will be working only with latest versions, they should not use the versioned flag on the filesystem instance, even when the bucket is indeed versioned? We should document this.

Reply:
Perhaps in the docstring of S3FileSystem at the argument version_aware?